File: solutions/php/0278-first-bad-version.php
<?php

/* The isBadVersion API is defined in the parent class VersionControl.
      public function isBadVersion($version){} */

class Solution extends VersionControl
{
    /**
     * @param Integer $n
     * @return Integer
     */
    public function firstBadVersion(int $n): int
    {
        $left = 1;
        $right = $n;
        $firstBad = $n;

        while ($left <= $right) {
            $currentVersion = intval(($left + $right) / 2);

            if ($this->isBadVersion($currentVersion)) {
                $firstBad = $currentVersion;
                $right = $currentVersion - 1;
            } else {
                $left = $currentVersion + 1;
            }
        }

        return $firstBad;
    }
}

File: solutions/go/0278-first-bad-version.go
package main

/**
 * Forward declaration of isBadVersion API.
 * @param   version   your guess about first bad version
 * @return 	 	      true if current version is bad
 *			          false if current version is good
 * func isBadVersion(version int) bool;
 */

func firstBadVersion(n int) int {
	left, right, bad := 1, n, n

	for left <= right {
		current := (left + right) / 2

		if isBadVersion(current) {
			bad = current
			right = current - 1
		} else {
			left = current + 1
		}
	}

	return bad
}
